### PR TITLE
Use fmt machinery instead of io one

### PR DESCRIPTION
**Pull request by est31**
This is in preparation of a #![no_std] port

### CODE REVIEW

**Comment by est31** [comment]
#![no_std] is needed for https://github.com/gimli-rs/addr2line/issues/89 .

File: src/ast.rs [hunk 1]
```diff
@@ -6021,7 +6015,8 @@ where
             } else {
                 start
             };
-            ctx.write_all(&ctx.input[start..end])
+            let s = ::std::str::from_utf8(&ctx.input[start..end]).map_err(|_| fmt::Error)?;
```
Comment (est31) [inline src/ast.rs]: Is this `.map_err(|_| fmt::Error)` okay? I'm doing this in multiple places in this PR. I could also create a custom error type or use the error type from the the error module of this crate (extending it with `fmt::Error` support previously).

**Comment by fitzgen** [comment]
@est31 the motivation is that `core::fmt` exists and `core::io` does not, correct?

**Comment by est31** [comment]
@fitzgen exactly. We only use `std::io` in order to write stuff, so we can fully replace it with `std::fmt` usage, which exists inside `core`.

**Pull request merged**